src/log.rs

```diff
@@ -61,38 +60,94 @@ macro_rules! emit_event {
     };
 }
 
-pub trait LogExt<T> {
+pub trait LogExt<T, E>
+where
+    Self: std::marker::Sized,
+{
+    #[track_caller]
+    fn format_log(self, context: &Context, msg: Option<&str>) -> Result<T, E>;
+
     /// Emits a warning if the receiver contains an Err value.
     ///
-    /// Returns an [`Option<T>`] with the `Ok(_)` value, if any:
-    /// - You won't get any warnings about unused results but can still use the value if you need it
-    /// - This prevents the same warning from being printed to the log multiple times
-    ///
     /// Thanks to the [track_caller](https://blog.rust-lang.org/2020/08/27/Rust-1.46.0.html#track_caller)
     /// feature, the location of the caller is printed to the log, just like with the warn!() macro.
+    ///
+    /// Unfortunately, the track_caller feature does not work on async functions (as of Rust 1.50).
+    /// Once it is, you can add `#[track_caller]` to helper functions that use one of the log helpers here
+    /// so that the location of the caller can be seen in the log. (this won't work with the macros,
+    /// like warn!(), since the file!() and line!() macros don't work with track_caller)  
+    /// See https://github.com/rust-lang/rust/issues/78840 for progress on this.
+    #[track_caller]
+    fn log_err(self, context: &Context, msg: &str) -> Result<T, E> {
+        self.format_log(context, Some(msg))
+    }
+
+    /// Emits a warning if the receiver contains an Err value and returns an [`Option<T>`].
+    ///
+    /// Example:
+    /// ```text
+    /// if let Err(e) = do_something() {
+    ///     warn!(context, "{:#}", e);
+    /// }
+    /// ```
+    /// is equivalent to:
+    /// ```text
+    /// do_something().ok_or_log(context);
+    /// ```
+    ///
+    /// For a note on the `track_caller` feature, see the doc comment on `log_err()`.
     #[track_caller]
-    fn log(self, context: &Context) -> Option<T>;
+    fn ok_or_log(self, context: &Context) -> Option<T> {
+        self.format_log(context, None).ok()
+    }
+
+    /// Like `ok_or_log()`, but you can pass an extra message that is prepended in the log.
+    ///
+    /// Example:
+    /// ```text
+    /// if let Err(e) = do_something() {
+    ///     warn!(context, "Something went wrong: {:#}", e);
+    /// }
+    /// ```
+    /// is equivalent to:
+    /// ```text
+    /// do_something().ok_or_log_msg(context, "Something went wrong");
+    /// ```
+    /// and is also equivalent to:
+    /// ```text
+    /// use anyhow::Context as _;
+    /// do_something().context("Something went wrong").ok_or_log(context);
+    /// ```
+    ///
+    /// For a note on the `track_caller` feature, see the doc comment on `log_err()`.
+    #[track_caller]
+    fn ok_or_log_msg(self, context: &Context, msg: &'static str) -> Option<T> {
+        self.format_log(context, Some(msg)).ok()
+    }
 }
 
-impl<T> LogExt<T> for anyhow::Result<T> {
+impl<T: Default, E: std::fmt::Display> LogExt<T, E> for Result<T, E> {
```

Comment:
Oh, I assumed you would need two impl blocks, the one you had `for anyhow::Result` and this one for `E: Display`.  I didn't realise that you can also use the alternate formatter on `E: Display` without any harm.  Which is to say that the only reason I introduced `format_log()` was because I thought you needed two impl blocks and that would reduce the code duplication for them.  Given that you don't I'm not sure `format_log` adds much to the party other than (a little) more complexity, you can just use the `log_err()` like you did before.  Apologies for this red herring.

Reply:
I think we still need `format_log()` as we can't use `.context()` anymore now. 
And if `log_err()` takes a `msg` param, then we can't implement `ok_or_log()` by calling through to `log_err()` (except by treating "" as a special value that indicates no message; what we could do, of course).
While, if `log_err()` doesn't take a `msg` param, then we can't implement `ok_or_log_msg()` as we can't add context without the `.context()` method (except by sth. like `.map_err(|e| format_err!("{}: {:#}", msg, e)`, which would be the third possible solution).

Reply:
oh, fair enough.  Some more bikeshedding would argue that `format_log` is a bit misleading since it does more than format.  I'd expect something to call format_log to return some kind of string and have no side effects.  anyway, maybe call it something like `inner_log_err()` or `log_err_opt_msg()` for want of better names.  but it doesn't matter a lot i guess

Reply:
I'll call it `log_err_inner()` (I think that `inner` is more often the suffix than the prefix in our codebase :)